src/richie/apps/core/fields/duration.py

```diff
@@ -0,0 +1,275 @@
+    """
+
+    def to_python(self, value):
+        """Return the raw value to avoid the charfield behavior that forces the value to str."""
+        return value
+
+
+class CompositeDurationField(models.CharField):
+    """
+    A custom charfield to store a duration in database as an integer and an interval choice.
+    """
+
+    description = _("Define a duration by intervals")
+
+    error_message = _("%(value)s is not a valid choice for an interval of time.")
+
+    def __init__(self, *args, **kwargs):
+        """
+        Record the two additional field attributes `intervals` and `default_interval` in the
+        instance.
+        """
+        self.intervals = kwargs.pop("intervals", None)
+        self.default_interval = kwargs.pop("default_interval", None)
+        super().__init__(*args, **kwargs)
+
+    def deconstruct(self):
+        """Return enough information to recreate the field as a 4-tuple."""
+        name, path, args, kwargs = super().deconstruct()
+        kwargs["intervals"] = self.intervals
+        kwargs["default_interval"] = self.default_interval
+        return name, path, args, kwargs
+
+    def _check_intervals_attribute(self):
+        """Check that `intervals` is well configured."""
+        if self.intervals is None:
+            return [
+                checks.Error(
+                    "CompositeDuration fields must define an 'intervals' attribute.",
+                    obj=self,
+                    id="fields.E1020",
+                )
+            ]
+        if not (
+            isinstance(self.intervals, dict)
+            and all(
+                [isinstance(c, tuple) and len(c) == 2 for c in self.intervals.values()]
+            )
+        ):
+            return [
+                checks.Error(
+                    "'intervals' must be a dictionary of tuples.",
+                    obj=self,
+                    id="fields.E1021",
+                )
+            ]
+
+        return []
+
+    def _check_default_interval_attribute(self):
+        """Check that `default_interval` is well configured."""
+        if self.default_interval and self.default_interval not in self.intervals:
+            return [
+                checks.Error(
+                    "'{:s}' is not a valid default interval.".format(
+                        self.default_interval
+                    ),
+                    obj=self,
+                    id="fields.E1022",
+                )
+            ]
+
+        return []
+
+    def check(self, **kwargs):
+        """Add checks on intervals to the field's checks."""
+        return (
+            super().check(**kwargs)
+            + self._check_intervals_attribute()
+            + self._check_default_interval_attribute()
+        )
+
+    @staticmethod
+    def from_db_value(value, *_args):
+        """Convert a database value to a list value."""
+        if not value:
+            return None
+        duration, interval = value.split("|")
```

Comment:
I was assuming that the content of the database is necessarily conform to what i wrote... I'm not sure what to do better than break in case the data is corrupted?

Reply:
Raise a `DataError` or an `IntegrityError`?